broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionLeaderElection.java

```diff
@@ -81,35 +82,75 @@ public void start(ServiceStartContext startContext) {
 
     LOG.info("Creating leader election for partition {} in node {}", partitionId, memberId);
 
+    startFuture = new CompletableActorFuture();
+
+    startContext.getScheduler().submitActor(this);
+    startContext.async(startFuture, true);
+  }
+
+  @Override
+  protected void onActorStarted() {
+    initialize();
+  }
+
+  private void initialize() {
     final CompletableFuture<LeaderElection<String>> leaderElectionCompletableFuture =
         atomix
             .<String>leaderElectionBuilder(DistributedLogstreamName.getPartitionKey(partitionId))
             .withProtocol(PROTOCOL)
             .buildAsync();
-
-    final CompletableActorFuture startFuture = new CompletableActorFuture();
-    leaderElectionCompletableFuture.thenAccept(
-        e -> {
-          election = e;
-          election.run(memberId);
-          startContext.getScheduler().submitActor(this).join();
-          startFuture.complete(null);
+    leaderElectionCompletableFuture.whenComplete(
+        (leaderElection, error) -> {
+          if (error == null) {
+            election = leaderElection;
+            actor.call(
+                () -> {
+                  final CompletableActorFuture<Void> joinFuture = new CompletableActorFuture<>();
+                  actor.runUntilDone(() -> tryJoinElection(joinFuture));
+                  actor.runOnCompletion(
+                      joinFuture, (r, e) -> actor.runUntilDone(this::tryAddListener));
+                });
+          } else {
+            LOG.debug(
+                "Couldn't create leader election for partition {}, retrying.", partitionId, error);
+            actor.call(this::initialize);
+          }
         });
+  }
 
-    startContext.async(startFuture, true);
+  private void tryJoinElection(CompletableActorFuture<Void> joinFuture) {
+    try {
+      if (canBecomeLeader) {
+        election.run(memberId);
+      }
+      joinFuture.complete(null);
+      actor.done();
```

Comment:
It's a little confusing for me that it calls `actor.done` and `actor.yield` outside of a `actor.runUntilDone`. How about this method returns a true/false if it joined, and wherever you call it in a `runUntilDone` you can check the result and decide whether to yield or not? Or a simpler solution, if these methods (`tryAddListener` and `tryJoinElection` are only used in `runUntilDone`, why not move the `runUntilDone` call in them? That way it's clearer that you **have** to call `done` or `yield`.